engine/evd.ml

```diff
@@ -205,21 +205,26 @@ module Store = Store.Make ()
 
 let string_of_existential evk = "?X" ^ string_of_int (Evar.repr evk)
 
-type evar_body =
-  | Evar_empty
-  | Evar_defined of constr
+type defined = [ `defined ]
+type undefined = [ `undefined ]
```

Comment:
Is there any reason for not making `defined` and `undefined` empty types?
```suggestion
type defined
type undefined
```
I've tested this on my machine and this seem to work.

Reply:
that's abstract not empty, and it only works because so far no code looks like `let get_body = function Evar_defined c -> c` outside evd.ml

Reply:
Indeed, it's weaker and not enough for the general elimination rule.